gnd/src/main/java/com/google/android/gnd/persistence/remote/firestore/schema/ObservationConverter.java

```diff
@@ -54,32 +56,61 @@ static Observation toObservation(Feature feature, DocumentSnapshot snapshot)
         .setProject(feature.getProject())
         .setFeature(feature)
         .setForm(form)
-        .setResponses(toResponseMap(doc.getResponses()))
+        .setResponses(toResponseMap(snapshot.getId(), form, doc.getResponses()))
         .setCreated(AuditInfoConverter.toAuditInfo(created))
         .setLastModified(AuditInfoConverter.toAuditInfo(lastModified))
         .build();
   }
 
-  private static ResponseMap toResponseMap(@Nullable Map<String, Object> docResponses) {
+  private static ResponseMap toResponseMap(
+      String observationId, Form form, @Nullable Map<String, Object> docResponses) {
     ResponseMap.Builder responses = ResponseMap.builder();
     if (docResponses == null) {
       return responses.build();
     }
     for (String fieldId : docResponses.keySet()) {
-      Object obj = docResponses.get(fieldId);
-      if (obj instanceof String) {
-        TextResponse.fromString(((String) obj).trim())
-            .ifPresent(r -> responses.putResponse(fieldId, r));
-        // TODO(#23): Add support for number fields:
-        // } else if (obj instanceof Float) {
-        //   responses.put(key, new NumericResponse((Float) obj));
-      } else if (obj instanceof List) {
-        MultipleChoiceResponse.fromList((List<String>) obj)
-            .ifPresent(r -> responses.putResponse(fieldId, r));
-      } else {
-        Timber.e("Unsupported obj in db: %s", obj.getClass().getName());
+      try {
+        Object obj = docResponses.get(fieldId);
+        putResponse(fieldId, form, obj, responses);
+      } catch (DataStoreException e) {
+        Timber.d("Field " + fieldId + "in remote db in observation " + observationId + ": " + e);
       }
     }
     return responses.build();
   }
+
```

Comment:
many of these classes are too low level to not have 100% unit test coverage. LGTM in general but strongly recommend having good unit test coverage here. In future across release if there are changes in this section, one will need to be quite a bit careful around backward compatibility etc.

Reply:
@atalekar Very good points. We started out working under the assumption that everything we're building is a throw-away prototype/demo, so we started without test coverage. The initial assumption probably no longer holds, but our 20% team is so small that adding tests retroactively wasn't feasible. That said, I agree we probably shouldn't kick the can further down the road. I'll try to add some tests for this class!

Reply:
Tests still a WIP, but added basic skeleton to get started. I'll flesh them out tomorrow, but please feel free to take a look at the last few commits now if you have a cycle or two. Thanks!

Reply:
lgtm - thanks a lot!